doc/source/ray-air/examples/pytorch_resnet_finetune.ipynb

```diff
@@ -0,0 +1,1152 @@
+    "- Use {meth}`iter_torch_batches <ray.data.Dataset.iter_torch_batches>` to iterate the datasets with automatic tensor batching and device placement. If you need a more flexible customized batching function, please refer to our lower-level {meth}`iter_batches <ray.data.Dataset.iter_batches>` API.\n",
+    "\n",
+    "**2. Prepare your model**:\n",
+    "- {meth}`train.torch.prepare_model() <ray.train.torch.prepare_model>` will prepares the model for distributed training. Under the hood, it converts your torch model to `DistributedDataParallel` model, which will synchronize its weights across all workers.\n",
+    "\n",
+    "**3. Report metrics and checkpoint**:\n",
+    "- {meth}`session.report() <ray.air.session.report>` will report metrics and checkpoints to Ray AIR.\n",
+    "- Saving checkpoints through {meth}`session.report(metrics, checkpoint=...) <ray.air.session.report>` will automatically [upload checkpoints to cloud storage](tune-cloud-checkpointing) (if configured), and allow you to easily enable Ray AIR worker fault tolerance in the future.\n",
+    "- The best checkpoints will be saved according to the specified `checkpoint_score_attribute` in {class}`CheckpointConfig <ray.air.config.CheckpointConfig>`. Here we only save the best model with highest validation accuracy."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 11,
+   "metadata": {
+    "id": "5pPf2kqKrLT2"
+   },
+   "outputs": [],
+   "source": [
+    "import ray.train as train\n",
+    "from ray.air import session\n",
+    "from ray.train.torch import TorchCheckpoint\n",
+    "\n",
+    "\n",
+    "def evaluate(logits, labels):\n",
+    "    _, preds = torch.max(logits, 1)\n",
+    "    corrects = torch.sum(preds == labels).item()\n",
+    "    return corrects\n",
+    "\n",
+    "\n",
+    "def train_loop_per_worker(configs):\n",
+    "    warnings.filterwarnings(\"ignore\")\n",
+    "    # Prepare dataloader for each worker\n",
+    "    datasets = dict()\n",
+    "    datasets[\"train\"] = session.get_dataset_shard(\"train\")\n",
+    "    datasets[\"val\"] = session.get_dataset_shard(\"val\")\n",
+    "\n",
+    "    # Calculate the batch size for a single worker\n",
+    "    worker_batch_size = configs[\"batch_size\"] // session.get_world_size()\n",
+    "\n",
+    "    device = train.torch.get_device()\n",
+    "\n",
+    "    # Prepare DDP Model, optimizer, and loss function\n",
+    "    model = initialize_model()  # [TODO]\n",
+    "\n",
+    "    model = train.torch.prepare_model(model)\n",
+    "\n",
+    "    optimizer = optim.SGD(\n",
+    "        model.parameters(), lr=configs[\"lr\"], momentum=configs[\"momentum\"]\n",
+    "    )\n",
+    "    criterion = nn.CrossEntropyLoss()\n",
+    "\n",
+    "    # Start training loops\n",
+    "    for epoch in range(configs[\"num_epochs\"]):\n",
+    "        # Each epoch has a training and validation phase\n",
+    "        for phase in [\"train\", \"val\"]:\n",
```

Comment:
how is validation supposed to work here? I think currently each worker does validation independently right? since the validation dataset is not split.

Reply:
I think validation should only be run on rank 0